File: src/py/bbctrl/LCD.py
################################################################################
#                                                                              #
#                 This file is part of the Buildbotics firmware.               #
#                                                                              #
#        Copyright (c) 2015 - 2020, Buildbotics LLC, All rights reserved.      #
#                                                                              #
#         This Source describes Open Hardware and is licensed under the        #
#                                 CERN-OHL-S v2.                               #
#                                                                              #
#         You may redistribute and modify this Source and make products        #
#    using it under the terms of the CERN-OHL-S v2 (https:/cern.ch/cern-ohl).  #
#           This Source is distributed WITHOUT ANY EXPRESS OR IMPLIED          #
#    WARRANTY, INCLUDING OF MERCHANTABILITY, SATISFACTORY QUALITY AND FITNESS  #
#     FOR A PARTICULAR PURPOSE. Please see the CERN-OHL-S v2 for applicable    #
#                                  conditions.                                 #
#                                                                              #
#                Source location: https://github.com/buildbotics               #
#                                                                              #
#      As per CERN-OHL-S v2 section 4, should You produce hardware based on    #
#    these sources, You must maintain the Source Location clearly visible on   #
#    the external case of the CNC Controller or other product you make using   #
#                                  this Source.                                #
#                                                                              #
#                For more information, email info@buildbotics.com              #
#                                                                              #
################################################################################

import lcd
import atexit


class LCDPage:
    def __init__(self, lcd, text = None):
        self.lcd = lcd
        self.data = lcd.new_screen()

        if text is not None:
            self.text(text, (lcd.width - len(text)) // 2, 1)


    def activate(self): pass
    def deactivate(self): pass


    def put(self, c, x, y):
        y += x // self.lcd.width
        x %= self.lcd.width
        y %= self.lcd.height

        if self.data[x][y] != c:
            self.data[x][y] = c
            if self == self.lcd.page: self.lcd.update()


    def text(self, s, x, y):
        for c in s:
            self.put(c, x, y)
            x += 1


    def clear(self):
        self.data = self.lcd.new_screen()
        self.lcd.redraw = True


    def shift_left(self): pass
    def shift_right(self): pass
    def shift_up(self): pass
    def shift_down(self): pass


class LCD:
    def __init__(self, ctrl):
        self.ctrl = ctrl
        self.log = ctrl.log.get('LCD')

        self.addrs = self.ctrl.args.lcd_addr
        self.addr = self.addrs[0]
        self.addr_num = 0

        self.width = 20
        self.height = 4
        self.lcd = None
        self.timeout = None
        self.reset = False
        self.page = None
        self.pages = []
        self.current_page = 0
        self.screen = self.new_screen()
        self.set_message('Loading...')

        self._redraw(False)
        if not ctrl.args.demo: atexit.register(self.goodbye)


    def set_message(self, msg):
        try:
            self.load_page(LCDPage(self, msg))
            self._update()
        except IOError as e:
            self.log.warning('LCD communication failed: %s' % e)


    def new_screen(self):
        return [[' ' for y in range(self.height)] for x in range(self.width)]


    def new_page(self): return LCDPage(self)
    def add_page(self, page): self.pages.append(page)


    def add_new_page(self, page = None):
        if page is None: page = self.new_page()
        page.id = len(self.pages)
        self.add_page(page)
        return page


    def load_page(self, page):
        if self.page != page:
            if self.page is not None: self.page.deactivate()
            page.activate()
            self.page = page
            self.redraw = True
            self.update()


    def set_current_page(self, current_page):
        self.current_page = current_page % len(self.pages)
        self.load_page(self.pages[self.current_page])


    def page_up(self): pass
    def page_down(self): pass
    def page_right(self): self.set_current_page(self.current_page + 1)
    def page_left(self): self.set_current_page(self.current_page - 1)


    def update(self):
        if self.timeout is None:
            self.timeout = self.ctrl.ioloop.call_later(0.25, self._update)


    def _redraw(self, now = True):
        if now:
            self.redraw = True
            self.update()
        self.redraw_timer = self.ctrl.ioloop.call_later(5, self._redraw)


    def _update(self):
        self.timeout = None

        try:
            if self.lcd is None:
                self.lcd = lcd.LCD(self.ctrl.i2c, self.addr, self.height,
                                   self.width)

            if self.reset:
                self.lcd.reset()
                self.redraw = True
                self.reset = False

            cursorX, cursorY = -1, -1

            for y in range(self.height):
                for x in range(self.width):
                    c = self.page.data[x][y]

                    if self.redraw or self.screen[x][y] != c:
                        if cursorX != x or cursorY != y:
                            self.lcd.goto(x, y)
                            cursorX, cursorY = x, y

                        self.lcd.put_char(c)
                        cursorX += 1
                        self.screen[x][y] = c

            self.redraw = False

        except IOError as e:
            # Try next address
            self.addr_num += 1
            if len(self.addrs) <= self.addr_num: self.addr_num = 0
            self.addr = self.addrs[self.addr_num]
            self.lcd = None

            self.log.warning('LCD communication failed, ' +
                        'retrying on address 0x%02x: %s' % (self.addr, e))

            self.reset = True
            self.timeout = self.ctrl.ioloop.call_later(1, self._update)


    def goodbye(self, message = ''):
        if self.timeout:
            self.ctrl.ioloop.remove_timeout(self.timeout)
            self.timeout = None

        if self.redraw_timer:
            self.ctrl.ioloop.remove_timeout(self.redraw_timer)
            self.redraw_timer = None

        if self.lcd is not None: self.set_message(message)

File: src/avr/src/lcd.c
/******************************************************************************\

                  This file is part of the Buildbotics firmware.

         Copyright (c) 2015 - 2021, Buildbotics LLC, All rights reserved.

          This Source describes Open Hardware and is licensed under the
                                  CERN-OHL-S v2.

          You may redistribute and modify this Source and make products
     using it under the terms of the CERN-OHL-S v2 (https:/cern.ch/cern-ohl).
            This Source is distributed WITHOUT ANY EXPRESS OR IMPLIED
     WARRANTY, INCLUDING OF MERCHANTABILITY, SATISFACTORY QUALITY AND FITNESS
      FOR A PARTICULAR PURPOSE. Please see the CERN-OHL-S v2 for applicable
                                   conditions.

                 Source location: https://github.com/buildbotics

       As per CERN-OHL-S v2 section 4, should You produce hardware based on
     these sources, You must maintain the Source Location clearly visible on
     the external case of the CNC Controller or other product you make using
                                   this Source.

                 For more information, email info@buildbotics.com

\******************************************************************************/

#include "lcd.h"
#include "oled.h"
#include "rtc.h"
#include "hardware.h"
#include "command.h"

#include <avr/io.h>
#include <avr/wdt.h>
#include <util/delay.h>

#include <stdbool.h>


void lcd_init(uint8_t addr) {
  // Enable I2C master
  TWIC.MASTER.BAUD = F_CPU / 2 / 100000 - 5; // 100 KHz
  TWIC.MASTER.CTRLA = TWI_MASTER_ENABLE_bm;
  TWIC.MASTER.CTRLB = TWI_MASTER_TIMEOUT_DISABLED_gc;
  TWIC.MASTER.STATUS = TWI_MASTER_BUSSTATE_IDLE_gc;

  _delay_ms(50);
  lcd_nibble(addr, 3 << 4); // Home
  _delay_ms(50);
  lcd_nibble(addr, 3 << 4); // Home
  _delay_ms(50);
  lcd_nibble(addr, 3 << 4); // Home
  lcd_nibble(addr, 2 << 4); // 4-bit

  lcd_write(addr,
            LCD_FUNCTION_SET | LCD_2_LINE | LCD_5x8_DOTS | LCD_4_BIT_MODE, 0);
  lcd_write(addr, LCD_DISPLAY_CONTROL | LCD_DISPLAY_ON, 0);
  lcd_write(addr, LCD_ENTRY_MODE_SET | LCD_ENTRY_SHIFT_INC, 0);

  lcd_write(addr, LCD_CLEAR_DISPLAY, 0);
  lcd_write(addr, LCD_RETURN_HOME, 0);
}


static void _master_wait() {
#ifdef __AVR__
  while (!(TWIC.MASTER.STATUS & TWI_MASTER_WIF_bm)) continue;
#endif
}


static void _write_i2c(uint8_t addr, uint8_t data) {
  data |= BACKLIGHT_BIT;

  TWIC.MASTER.ADDR = addr << 1;
  _master_wait();

  TWIC.MASTER.DATA = data;
  _master_wait();

  TWIC.MASTER.CTRLC = TWI_MASTER_CMD_STOP_gc;

  _delay_us(100);
}


void lcd_nibble(uint8_t addr, uint8_t data) {
  _write_i2c(addr, data);
  _write_i2c(addr, data | ENABLE_BIT);
  _delay_us(500);
  _write_i2c(addr, data & ~ENABLE_BIT);
  _delay_us(100);
}


void lcd_write(uint8_t addr, uint8_t cmd, uint8_t flags) {
  lcd_nibble(addr, flags | (cmd & 0xf0));
  lcd_nibble(addr, flags | ((cmd << 4) & 0xf0));
}


void lcd_goto(uint8_t addr, uint8_t x, uint8_t y) {
  static uint8_t row[] = {0, 64, 20, 84};
  lcd_write(addr, LCD_SET_DDRAM_ADDR | (row[y] + x), 0);
}


void lcd_putchar(uint8_t addr, uint8_t c) {
  lcd_write(addr, c, REG_SELECT_BIT);
}


void lcd_pgmstr(uint8_t addr, const char *s) {
  while (true) {
    char c = pgm_read_byte(s++);
    if (!c) break;
    lcd_putchar(addr, c);
  }
}


void _splash(uint8_t addr) {
  lcd_init(addr);
  lcd_goto(addr, 0, 0);
  lcd_pgmstr(addr, PSTR("Controller booting"));
  lcd_goto(addr, 0, 1);
  lcd_pgmstr(addr, PSTR("Please wait..."));
  lcd_goto(addr, 0, 3);
  lcd_pgmstr(addr, PSTR("(c) Buildbotics LLC"));
}


void lcd_splash() {
  wdt_disable();
  oled_splash();
  _splash(0x27);
  _splash(0x3f);
  wdt_enable(WDTO_250MS);
}


void lcd_rtc_callback() {
  // Display the splash if we haven't gotten any commands in 1sec since boot
  if (!command_is_active() && rtc_get_time() == 1000)
    lcd_splash();
}
